### DISCUSSION TITLE
How does "gem install" know to use the openssl 1.1 dir created by ruby-build?

### DISCUSSION THREAD

**Discussion by bughit** [rbenv/ruby-build#2414] [General]
In a dump of a failing mysql2 gem install I could see that its using the correct openssl 1.1 dir that was built by ruby-build:

`Using --with-openssl-dir=/home/user1/.rbenv/versions/2.6.10/openssl`

How does it know to do that? Is it simply the presence of openssl dir in the ruby dir? Or is there some config file that specifies it?

Ultimately I'd like to know if there's any possibility of this mechanism failing and a gem build using the wrong (Ubuntu 24.04 default 3.0) openssl dir.

**Comment by mislav** [comment]
Hi, any ruby program can inspect the flags that the current ruby is built with by importing the `rbconfig` package:
```
ruby -r rbconfig -e 'puts RbConfig::CONFIG["configure_args"]'
```
I'm not sure if mysql2 is automatically doing this during its gem compilation. (I'm looking at its [extconf.rb](https://github.com/brianmario/mysql2/blob/b2b6d7b1498316058ce00c993bcb98788a8d1a92/ext/mysql2/extconf.rb) but it doesn't look like it ever reads the `"configure_args"` key.) After reading its documentation, I would guess it doesn't do this itself, but gives instructions on how the user might do this themselves: https://github.com/brianmario/mysql2/blob/b2b6d7b1498316058ce00c993bcb98788a8d1a92/README.md?plain=1#L126-L132

So I would guess that the explicit `--with-openssl-dir` flag during mysql2 compilation comes from user configuration somewhere, such as bundler global/local configuration.

**Comment by bughit** [comment]
If its up to each gem to choose the right openssl-dir, this seems dangerous/unreliable, some may choose incorrectly. So hopefully it's not that.

Bundler is not involved in this test, its a `gem install` from the command-line. Could you give any further tips for figuring out how this works? I looked thought local configuration, its a new install/account so there wasn't a lot, didn't notice anything. Also not in the shell environment.

**Reply by mislav** [reply]
> If its up to each gem to choose the right openssl-dir, this seems dangerous/unreliable, some may choose incorrectly. So hopefully it's not that.

You're right; upon closer inspection it looks like the mysql2 gem automatically respects the openssl directory that the ruby interpreter was configured with: https://github.com/brianmario/mysql2/blob/b2b6d7b1498316058ce00c993bcb98788a8d1a92/ext/mysql2/extconf.rb#L49-L60

Basically, mysql2 checks whether the ruby has been configured with `--with-openssl-dir`. It makes the equivalent of this check:
```
ruby -r mkmf -e 'p with_config("openssl-dir")'
```
and if it finds a value, it adds its "lib" dir to LDFLAGS while compiling the mysql2 gem C extensions.

Alternatively, it looks like the gem build system may look up the [CONFIGURE_ARGS environment variable](https://github.com/ruby/ruby/blob/master/lib/mkmf.rb#L113) set at the time the `gem install` process runs.
